### DISCUSSION TITLE
Subroutines to transpose from X->Z and Z->X

### DISCUSSION THREAD

**Discussion by ygsvanshu** [xcompact3d/2decomp-fft#353] [Q&A] [answered]
Hi all!

After having used the library for incompressible fluid flow solvers, I think it's a very nice tool for computing FFTs on distributed memory systems! I was wondering if there is any interest towards including subroutines that transpose directly from X pencils -> Z pencils and vice versa? I know this can be achieved with two transforms X -> Y and Y -> Z but I guess that a direct transpose would probably offer a big performance improvement, especially when transforms need to be computed repeatedly.

**Comment by mathrack** [comment, marked as answer]
Hello,

Additional transpose subroutines could be used. The x->z and z->x transpose could be similar to the ones in https://github.com/PhysicsofFluids/AFiD.

However, when a case needs a lot of transpose x->z, often one could change the reference frame to avoid such transpose operations.

This new functionality is not under development currently. If you want to add it, feel free to let us know, we can help you to elaborate a pull request.

**Reply by ygsvanshu** [reply]
Thanks for your response!
Indeed I'm aware of the implementation from the Physics of Fluids group, I finished my PhD there and used the code AFiD (and modified versions) extensively. The transforms in X->Z and Z->X lack several features. For example, these were implemented only for complex type using MPI_Alltoallw instead of MPI_Alltoallv, without the option for "even" transposes where the dimension of the array is perfectly divisible by the number of processors along the direction, and also without the "padded" option. Therefore I thought it was good to simply inquire if there was any interest in official support for this feature. In the future, I might attempt to add this feature if I find some spare time. If I find that my attempts have been successful I might get back about a pull request  :)

**Reply by mathrack** [reply]
Overall, the transpose subroutines could use some upgrade. I think we could keep the current frontend and offer various backends under the hood : nccl / MPI_ALLTOALL / MPI_ALLTOALLV / mpi_send_recv

This will probably be part of the next major version of 2decomp. Feel free to drop a wishlist / suggestions for the next version in the issues section !
